Beta testing (#319)
UniversalRobots/Universal_Robots_ROS_Driver: ur_robot_driver/README.md

* Cartesian control using the joint-based robot interface

All of them can be used with corresponding ROS controllers.
More information about the controllers can be found in the cartesian ros control repository.

##### Cartesian pose streaming

Start the preloaded `pose_traj_controller`.
This controller allows ROS-level interpolation and streaming of cartesian poses to the robot.
It uses the newly created FollowCartesianTrajectory action.

##### Cartesian velocity streaming

Start the preloaded `twist_controller`.
This controller allows for streaming twist commands via the `twist_controller/command` topic to be
interpreted as cartesian velocity control commands by the robot.

##### Joint-based trajectory forwarding

Start the preloaded `forward_joint_traj_controller`.
This controllers allows for forwarding full trajectories to the robot, to be interpolated and
executed completely by the internal robot control.
While it uses the same FollowJointTrajectory interface as the previously existing trajectory
control, during execution no control commands other than a keepalive signal are sent to the robot.
A cancellation of the trajectory is still possible.

> Of special not for the forwarded trajectory is that, currently, blending is activated on the robot.

Reviewer: Maybe rephrase a little.
```suggestion
Note that blending is activated for forwarded trajectories by default.
```

This produces smoother trajectories, but causes target points other than the last to not
necessarily be reached.
One of the objectives of this beta test is to gather feedback for this behaviour and look into
feasible ways to control this blending radius, that has to directly corresponding field in the ROS

> trajectory interface definition, to be set based on the commanded trajectory.

Reviewer 2: What about using the `path_tolerance` in some way?

Author: Yes, the path_tolerance should most likely be the basis for this - we do have to move from joint based to position(/velocity/acceleration) tolerances to a cartesian radius, though.
As this transformation is dependent on the robot configuration, there are some decisions on complexity vs. accuracy to be made here - especially thinking about transparency for the person setting these path_tolerance values.

Reviewer 2: This is discussing the `forward_joint_traj_controller`is it not? Or all forwarding controllers?

Reviewer: That would concern both forwarding controllers. There are two types, one for joint-based trajectories and one for Cartesian-based trajectories.

Author: Yes, at least primarily. For the cartesian forwarding controller using the position error in the cartesian tolerance to calculate a good blend radius should be trivial.

##### Cartesian trajectory forwarding

Start the preloaded `forward_cartesian_traj_controller`.
Similar to the previous controller, this offers forwarding cartesian trajectories to the robot.
It is used in the same general structure and also utilizes blending.

##### Cartesian control using joint-based robot interface

> Start the preloaded `vendor_cartesian_traj_controller`.

Reviewer 2: What's the rationale for prefixing this with `vendor_`?

Before reading the explanatory text below, I sort-of assumed this would be a pure forwarding controller, but it seems that's the role of the `forward_cartesian_traj_controller`.

Reviewer: `vendor_ ` is an artifact from a different context. I would suggest dropping this and use the more verbose names:

`joint_based_cartesian_traj_controller` and `pose_based_cartesian_traj_controller`.
